cpp/src/arrow/util/cache_internal.h

```diff
@@ -0,0 +1,232 @@
+      items_.splice(items_.begin(), items_, list_it);
+      list_it->value = std::forward<V>(value);
+      return {false, &list_it->value};
+    }
+  }
+
+ private:
+  struct Item {
+    // Pointer to the key inside the unordered_map
+    const Key* key;
+    Value value;
+  };
+  using List = std::list<Item>;
+  using ListIt = typename List::iterator;
+
+  const int32_t capacity_;
+  // In most to least recently used order
+  std::list<Item> items_;
+  std::unordered_map<Key, ListIt> map_;
+};
+
+namespace detail {
+
+template <typename Key, typename Value, typename Cache, typename Func>
+struct ThreadSafeMemoizer {
+  using RetType = Value;
+
+  template <typename F>
+  ThreadSafeMemoizer(F&& func, int32_t cache_capacity)
+      : mutex_(new std::mutex), func_(std::forward<F>(func)), cache_(cache_capacity) {}
+
+  // The memoizer can't return a pointer to the cached value, because
+  // the cache entry may be evicted by another thread.
+
+  Value operator()(const Key& key) {
+    std::unique_lock<std::mutex> lock(*mutex_);
+    const Value* value_ptr;
+    value_ptr = cache_.Find(key);
+    if (ARROW_PREDICT_TRUE(value_ptr != NULLPTR)) {
+      return *value_ptr;
+    }
+    lock.unlock();
+    Value v = func_(key);
+    lock.lock();
+    return *cache_.Replace(key, std::move(v)).second;
+  }
+
+ private:
+  std::unique_ptr<std::mutex> mutex_;
+  Func func_;
+  Cache cache_;
+};
+
+template <typename Key, typename Value, typename Cache, typename Func>
+struct ThreadUnsafeMemoizer {
+  using RetType = const Value&;
+
+  template <typename F>
+  ThreadUnsafeMemoizer(F&& func, int32_t cache_capacity)
+      : /*mutex_(new std::mutex), */ func_(std::forward<F>(func)),
+        cache_(cache_capacity) {}
+
+  const Value& operator()(const Key& key) {
+    const Value* value_ptr;
+    value_ptr = cache_.Find(key);
+    if (ARROW_PREDICT_TRUE(value_ptr != NULLPTR)) {
+      return *value_ptr;
+    }
+    return *cache_.Replace(key, func_(key)).second;
+  }
+
+ private:
+  Func func_;
+  Cache cache_;
+};
+
+// A copy-constructible callable wrapper, for std::function<>
+
+template <typename Callable, typename RetType = call_traits::return_type<Callable>,
+          typename Value = call_traits::argument_type<0, Callable>>
+struct SharedCallable {
+  explicit SharedCallable(Callable callable)
+      : callable_(std::make_shared<Callable>(std::move(callable))) {}
+
+  template <typename V>
+  RetType operator()(V&& v) {
+    return (*callable_)(std::forward<V>(v));
+  }
+
+ private:
+  const std::shared_ptr<Callable> callable_;
+};
+
+template <template <typename...> class Cache,
+          template <typename...> class MemoizerType = ThreadSafeMemoizer, typename Func,
+          typename Key = typename std::decay<call_traits::argument_type<0, Func>>::type,
+          typename Value = typename std::decay<call_traits::return_type<Func>>::type,
+          typename Memoizer = MemoizerType<Key, Value, Cache<Key, Value>, Func>,
```

Comment:
This allows a memoizer to wrap the instance of `Func` by reference, which I think will be surprising. I've assembled an alternative in https://github.com/pitrou/arrow/pull/7

Reply:
I took some of the suggestions in your PR (thanks!). However, I kept the ability to store a reference, as it makes testing easier.